midend/replaceSelectRange.h

```diff
@@ -0,0 +1,66 @@
+/*
+* Copyright 2020, MNK Labs & Consulting
+* http://mnkcg.com
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*/
+#ifndef _MIDEND_REPLACESELECTRANGE_H_
+#define _MIDEND_REPLACESELECTRANGE_H_
+
+#include <iostream>
+#include <iomanip>
+#include <vector>
+#include <utility>
+#include "../lib/gmputil.h"
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+
+namespace P4 {
+
+/**
+  Find the types to replace and insert them in the range-mask map.
+  Call pass after the flattenInterfaceStructs pass.
+  */
+class DoReplaceSelectRange : public Transform {
+P4::TypeMap* typeMap;
+
+ public:
+    explicit DoReplaceSelectRange(TypeMap* typeMap): typeMap(typeMap) {
+        setName("DoReplaceSelectRange");
+        CHECK_NULL(typeMap);
+    }
+    const IR::Node* postorder(IR::SelectCase* p) override;
+    bool checkRange(const IR::Range* range);
+    std::vector<const IR::Mask *> rangeToMasks(const IR::Range *);
+    std::vector<IR::Vector<IR::Expression>> cartesianAppend(
+               std::vector<IR::Vector<IR::Expression>> vecs,
+               std::vector<const IR::Mask *> masks);
+    std::vector<IR::Vector<IR::Expression>> cartesianAppend(
+               std::vector<IR::Vector<IR::Expression>> vecs,
+               const IR::Expression *e);
+};
+
+class ReplaceSelectRange final : public PassManager {
+ public:
+        ReplaceSelectRange(ReferenceMap* refMap, TypeMap* typeMap) {
+            passes.push_back(new TypeChecking(refMap, typeMap));
+            passes.push_back(new DoReplaceSelectRange(typeMap));
+            passes.push_back(new ClearTypeMap(typeMap));
```

Comment:
I don't think this is necessary.

Reply:
You are correct because the IR::Range nodes are removed and do not change type.